src/digraph.h

```diff
@@ -0,0 +1,284 @@
+      if (i >= nr_nodes()) {
+        throw LibsemigroupsException("get: first argument larger than "
+                                     "number of nodes - 1");
+      }
+      if (j >= _next_edge_pos[i]) {
+        throw LibsemigroupsException("get: second argument larger than "
+                                     "number of edges from node - 1");
+      }
+      return _recvec.get(i, j);
+    }
+
+    //! Add some nodes to the graph
+    //!
+    //! Adds \p nr nodes to \c this.
+    void inline add_nodes(size_t nr) {
+      _recvec.add_rows(nr);
+      for (size_t i = 0; i < nr; ++i) {
+        _next_edge_pos.push_back(0);
+      }
+      _has_scc = false;
+    }
+
+    //! Returns the number of nodes of \c this
+    //!
+    //! Returns the number of nodes of \c this
+    TIntType inline nr_nodes() {
+      return _recvec.nr_rows();
+    }
+
+    //! Add an edge to this digraph
+    //!
+    //! If \p i and \p j are nodes in \c this, then this function
+    //! adds an edge from \p i to \p j.
+    void inline add_edge(TIntType i, TIntType j) {
+      if (i >= nr_nodes()) {
+        throw LibsemigroupsException("add_edge: first argument larger than "
+                                     "number of nodes - 1");
+      }
+      if (j >= nr_nodes()) {
+        throw LibsemigroupsException("add_edge: second argument larger than "
+                                     "number of nodes - 1");
+      }
+      if (_next_edge_pos[i] >= _degree_bound) {
+        throw LibsemigroupsException("add_edge: adding an edge would increase "
+                                     "the degree past the degree bound");
+      }
+      set(i, _next_edge_pos[i], j);
+      ++_next_edge_pos[i];
+    }
+
+    //! Returns the number of edges of \c this
+    //!
+    //! Returns the total number of edges of \c this
+    TIntType inline nr_edges() {
+      size_t edges = 0;
+      for (size_t i : _next_edge_pos) {
+        edges += i;
+      }
+      return edges;
+    }
+
+    //! Increase the degree of \c this
+    //!
+    //! Increases the maximum degree of \c this by \p nr.
+    //! This function must be called before you attempt to add an edge
+    //! to a vertex and increase its degree past \c degree_bound.
+    void inline increase_degree(TIntType nr) {
+      _recvec.add_cols(nr);
+      _degree_bound += nr;
+    }
+
+    //! Returns a bound on the degree of this graph
+    //!
+    //! This function returns the current bound on the degree of this graph.
+    //! This is used to decide how many columns to store in the RecVec.
+    TIntType degree_bound() {
+      return _degree_bound;
+    }
+
+    //! Calculate the strongly connected components of \c this
+    //!
+    //! This function calculates the strongly connected components of \c this
+    //! using Gabow's algorithm
+    //! The implementation is strongly based on that in the Digraphs package
+    //! (https://github.com/gap-packages/Digraphs)
+    void gabow_scc() {
```

Comment:
I'd be tempted to make this private, otherwise, it will repeat the computation if called multiple times

Reply:
Done